Task: Enhance logging and improve dev workflow

jonrandahl opened the pull request:
This commit introduces structured logging using JSON format, implements better error handling, and updates gem dependencies.  It also adds pre-commit, post-commit, and pre-push git hooks, along with development environment configurations.

Applies to #156

## Changes

- Introduces structured logging with JSON format.
- Improves error handling and reporting, using Sentry for error tracking.
- Adds git hooks for pre-commit, post-commit, and pre-push actions to enforce code quality and build processes.
- Updates gem dependencies and adds new gems.
- Modifies configurations for development environments, such as adding a Procfile.dev.
- Removes the old action_controller_prometheus_subscriber and introduces action_dispatch_prometheus_subscriber.

## Impact

- Improves application observability through structured logging.
- Enhances code quality and consistency with added git hooks and rubocop configuration.
- Updates dependency management in development environment.
- Affects error reporting and monitoring.

jonrandahl committed:
refactor: Update boot info logging for Rails compatibility

- Changed the structure to support a new command module.
- Updated the method to use 'say' instead of 'puts' for output.
- Simplified parameters in the boot information method.

jonrandahl committed:
feat: add pre-commit and pre-push hooks

- Introduced a pre-commit hook to run Rubocop on staged Ruby files.
- Added a pre-push hook to execute tests if new or modified test files are present.
- Both hooks skip checks for specific branch names (e.g., rebase, production).
- Hooks provide feedback on linting and testing status with clear messages.

jonrandahl committed:
chore: Update method comment for clarity

- Added a rubocop disable comment for method length
- Improved readability of the code with clearer comments

jonrandahl committed:
chore: Update timezone handling in ActiveSupport

- Added deprecation warning for `to_time` behaviour change
- Configured `to_time_preserves_timezone` to use system timezone

jonrandahl committed:
fix: Correct spelling errors in error messages

- Fixed "retreive" to "retrieve" in multiple error logs.
- Improved clarity of logging for month, year, and quarter retrieval failures.

jonrandahl committed:
feat: Enhance logging and error handling

- Added URL helpers and translation support.
- Renamed `detailed_request_log` to `detailed_log_result`.
- Improved log structure with additional fields like request time and user agent.
- Enhanced message formatting for successful requests.
- Updated error logging based on response status codes.

jonrandahl committed:
refactor: adjusted helpers and models via rubocop

- Changed `tag(:br)` to `tag.br` for cleaner syntax.
- Updated initialisation of hash objects to use `ActiveSupport::HashWithIndifferentAccess`.
- Simplified block passing in methods by using shorthand syntax.
- Adjusted cache control headers formatting in production config.
- Replaced directory loading method with a more concise glob approach.
- Fixed test assertions to use consistent negation style.

jonrandahl committed:
feat: update dependencies and add new gems

- Upgraded Rails and related gems to version 8.0.1
- Added dotenv for environment variable management
- Included rubocop and rubocop-rails for code linting
- Removed unicorn from development group
- Updated various other gem versions for compatibility and improvements

jonrandahl committed:
feat: Add development environment configuration

- Introduced a new environment file for development.
- Set up API service URL and metrics port.
- Configured Sentry settings for the development stage.

jonrandahl committed:
feat: Add RuboCop Rails plugin

- Included the rubocop-rails plugin for better integration
- Updated configuration to enhance code quality checks

jonrandahl committed:
feat: adds post-commit hook

Adds a git hook to automatically run tests before committing.

 - Implements logic to skip the hook if using `--no-verify` or `--amend`.
 - Adds checks for new test files and runs them individually.
 - Runs all tests if no new test files are found.
 - Prevents commit if tests fail.

jonrandahl committed:
build: Allows commits with failing tests (WIP)

 - Skips running tests on the specific branch.
 - Adds logic to allow commits even if tests fail.

jonrandahl committed:
chore: refactors the old pre-push hook script

 - Removes the previous implementation of the pre-push hook
 - Replaces with docker image build & run test process

jonrandahl committed:
feat: Configures log level via environment variable

- Configures the application log level using the `LOG_LEVEL` environment variable.
 -  Allows dynamic adjustment of the log level without requiring code changes.
 -  Defaults to 'debug' in development and test, and 'info' in production.

jonrandahl committed:
refactor: Improves details polyfill initialisation

 - Prevents unnecessary initialisation when native details support is present.
 - Renames `NATIVE_DETAILS` to `HAS_NATIVE_DETAILS` for clarity.

jonrandahl committed:
chore: Corrects typo in post-commit hook

jonrandahl committed:
build: Updates rubocop configuration

 - Sets the target Ruby version to 3.3.
 - Increases the allowed complexity for methods.
 - Excludes test files from certain complexity checks.

jonrandahl committed:
fix: Updates API service URL and ignores env files

- Updates the API service URL to use a different port, likely to avoid conflicts or to align with a new service configuration.
- Prevents accidental committing of local environment files, improving security and consistency across environments.

jonrandahl committed:
refactor: updates to metrics collection for errors

 - Renames the `ActionControllerPrometheusSubscriber` to `ActionDispatchPrometheusSubscriber` and updates it to subscribe to `:action_dispatch` events.
 - Improves error handling and reporting in the `ApiPrometheusSubscriber` to provide more detailed metrics on API connection failures and service exceptions.
 - Enhances the `ApplicationPrometheusSubscriber` to include the error type, message, and status code in the metrics for internal application errors.

jonrandahl committed:
feat: Enhances application logging

 - Introduces a helper module to manage the request logging.
 - Adds request logging to multiple controllers.
 - Normalises log output across the application.

jonrandahl committed:
refactor: Adds labels to internal error counter

 - Adds `result` and `status` labels to the `internal_application_error` Prometheus counter to allow for better debugging.

jonrandahl committed:
feat: reconfigures Sentry for better error tracking

 - Adds environment tags to Sentry events for better filtering, enhanced error tracking and performance monitoring.
 - Sets up Sentry with API key, environment, and version.
 - Configures breadcrumbs, excluded exceptions, and sampling rates.

jonrandahl committed:
refactor: update Spring pre-loading

- Removes Spring pre-loading from the `bin/rails` and `bin/rake` files.
 - Simplifies the binstubs by removing the Spring pre-loading logic.
 - Removes the `bin/setup` and `bin/spring` files.

jonrandahl committed:
refactor: reconfigures application for Rack servers

- Implements conditional loading of environment variables using Dotenv for non-production environments.
- Logs specified environment variables for debugging purposes.

jonrandahl committed:
refactor: Improve readability of toggle function

 - Converts `var` to `const` for variables that are not reassigned.
 - Minor formatting changes for consistency.

jonrandahl committed:
refactor: Reorders asset includes

- Improves asset loading order
- Moves common styles import to earlier location

jonrandahl committed:
refactor: Improves form element styling

- Improves styling for form elements and report selection components
- Addresses inconsistencies in focus states and element scaling
- Enhances the display of report details within the selection component

jonrandahl committed:
feat: Improves report example image rendering

 - Enhances the display of report example images by using the `<picture>` element, allowing for different image sources based on screen size.

jonrandahl committed:
refactor: Reduce application dependencies

 - Removes unnecessary railties

jonrandahl committed:
fix: Improves internal error instrumentation

 - Enhances error reporting to include more contextual information such as message, status, type, cause, and backtrace.
 - Modifies the instrumentation process to only notify subscribers for errors with a status code of 500 or greater, preventing unnecessary notifications for non-critical issues.
 - Updates the logging mechanism to use the appropriate severity level (warn or error) based on the status code.

jonrandahl committed:
refactor: Enforces consistent rubocop configuration

- Improves rubocop configuration for consistency
- Adds an allowed pattern for long comment lines
- Uses double quotes for string literals within the `.rubocop.yml` file

jonrandahl committed:
build: Updates gems and adds dev tools

Updates gem dependencies and adds development tools to improve development workflow and stability:

 - Updates `json_rails_logger` and `lr_common_styles` gems to latest versions
 - Adds `dotenv` to allow use in the config file
 - Adds `puma` as the app server
 - Updates `sentry-ruby` to `sentry-rails`
 - Adds `foreman` and `ostruct` gems for development and testing
 - Adds `meta_request` for Rails development devtools panel
 - Removes `capybara_minitest_spec` gem

jonrandahl committed:
feat: Introduces structured logging

This commit introduces a centralised logging module to improve application observability and maintainability.

 - Replaces the existing `LoggingHelper` with a new `Log` module that provides structured logging using the Rails logger and JSON format.
 - Adds methods for logging messages at different levels (debug, info, warn, error) with configurable fields.
 - Ensures that requests are logged with relevant details such as status code, request time, and path.
 - Improves request logging by handling service API paths and request parameters.
 - Modifies the logging logic to use the appropriate log level based on the status code.
 - Removes the deprecated `LoggingHelper` file.

jonrandahl committed:
refactor: Improves request logging and error handling

 - Replaces `LoggingHelper` with `Log` for more streamlined logging.
 - Enhances error handling by providing more context in error messages.
 - Centralises logging functionality to improve code organisation.

jonrandahl committed:
refactor: Improves API client resilience

- Implements robust error handling in the API client, raising exceptions for non-200 status codes with detailed messages.
- Adds retry logic to the Faraday HTTP client to handle transient network issues.
- Improves instrumentation for API requests to track performance metrics.
- Improves logging for both successful and failed API requests, providing more context and detail.
- Simplifies the `ok?` method by raising an exception instead of returning false.
- Updates the `as_http_api` method to correctly handle API endpoints.

jonrandahl committed:
refactor: Use central log utility for request logging in controllers

 - Simplifies logging calls by using a dedicated logging utility
 - Ensures consistent logging across different parts of the application

jonrandahl committed:
refactor: Improves logging for report design workflow steps

 - Adds a more descriptive message to the log, including the step name and progress summary
 - Switches from using LoggingHelper to Log class
 - Removes a redundant rubocop disable comment

jonrandahl committed:
build: Increments minor version and resets patch

- Bumps the minor version to reflect new features and significant changes.
 - Resets the patch version to zero for a clean release.

jonrandahl committed:
feat: Adds development Procfile

Adds a Procfile for local development.

 - Defines commands to start both the web server and API server.
 - Configures environment variables for each process.

jonrandahl committed:
build: dockerlint dockerfile updates

jonrandahl committed:
docs: updates CHANGELOG

Improves application infrastructure and developer experience.

 - Implements structured logging for better integration with logging services.
 - Configures log level using environment variables.
 - Adds pre-commit and post-commit hooks for code quality and testing.
 - Introduces development environment configuration for easier setup.
 - Reconfigures Sentry for improved error tracking.
 - Updates dependencies and adds new gems for enhanced functionality.
 - Updates code styling via rubocop.

jonrandahl committed:
build: Updates reusable workflows to v2

- Updates the publish and deploy workflows to use version 2 of the reusable workflows.
- Aligns with the latest version for enhanced stability and features.

jonrandahl committed:
refactor: Improves log messages for better clarity and consistency

 - Updates log messages in controllers to indicate that a request is being made.
 - Improves report design log message to include the step name.

jonrandahl committed:
fix: Improves API request logging

 - Ensures that the URL is correctly formatted in debug logs.
 - Includes response body in error messages only when present.
 - Clarifies the success message for API requests.

jonrandahl committed:
refactor: Adds request logging to report manager

- Adds logging for report manager requests, including parameters and path
- Improves debugging and traceability of report requests
- Removes debugging statement in development environment, replaced by centralised logging

jonrandahl committed:
refactor: Reorders arguments in error logging

- Ensures the URL is provided before the exception when recording failed API connections.

jonrandahl committed:
refactor: update make run target

- Add `-d` to docker command to fork and run the image in the background

jonrandahl committed:
fix: Corrects ECR and API service URL

 - Ensures the ECR URL uses the correct AWS region.
 - Allows the API service to run on the local machine.
 - Adds a target to expose the version.

jonrandahl committed:
feat: Implements custom error pages

- Introduces custom error pages to provide a better user experience when encountering errors.
- Removes the default Rails error pages.
- Provides specific guidance and contact information for different error scenarios (400, 403, 404, 500).
- Includes a reference code (sentry_code) for support staff to identify the problem.

jonrandahl committed:
fix: Updates API service URL

- Updates the default API service URL to reflect a change in service naming and deployment.

jonrandahl committed:
feat: Enables automatic js-routes generation

- Automates the generation of js-routes files, ensuring they are up-to-date whenever the routes configuration changes.
- Adds a middleware to automatically update js-routes file in development.
- Configures js-routes to use a global `Routes` namespace.
- Ignores the generated js-routes files in `.gitignore`.

jonrandahl committed:
feat: Sets default contact email address

 - Specifies the default contact email address for Land Registry Data Services.

jonrandahl committed:
feat: Adds SENTRY_API_KEY check

- Implements a check for the presence of the SENTRY_API_KEY environment variable as per other LR apps
- Reports an error if the SENTRY_API_KEY is missing, aiding in debugging and ensuring proper configuration.

jonrandahl committed:
fix: Improves data validation and path handling

- Enhances data validation for log fields by using arrays for allowed values, ensuring consistency and accuracy in logging.
- Refines path parameter handling by ensuring that query parameters are only appended to the path if they exist. This prevents the creation of invalid or malformed paths.

jonrandahl committed:
chore: Updates `json_rails_logger` version

- Updates the json_rails_logger gem to the latest 2.0.6 version.

jonrandahl committed:
fix: Corrects docker run command arguments

 - Updates the `docker run` command arguments to align with Docker's expected syntax
 - Ensures the container runs in detached mode
 - Correctly publishes the specified port
 - Properly sets the environment variable for the API service URL

jonrandahl committed:
build: Ensures service is stopped before starting

- Splits chores into individual targets to maintain single responsibility
- Ensures the service is stopped before attempting to start it, providing a cleaner startup process.

jonrandahl committed:
chore: Updates SR Manager configuration for local testing

 - Updates the manager configuration to facilitate local testing.
 - Increases the connection timeout to accommodate local development environments.
 - Updates the API base URI to reflect the local setup.
 - Ignores automatically generated download logs.

jonrandahl committed:
fix: Renames error status variable

 - Updates the parameter name in the error rendering method to ensure correct value is passed to helper functions.

jonrandahl committed:
fix: Corrects error status assignment

 - Corrects a conditional check to properly determine if the status is a symbol before attempting to retrieve the corresponding HTTP status code.

jonrandahl committed:
refactor: Logs backtrace when debug logging is enabled

 - Ensures that backtraces are included in the error logs if the debug logging level is active, regardless of the Rails environment.

joescottdave commented on test/fixtures/conf/app.conf:
```diff
@@ -1,5 +1,11 @@
 # app configuration
 
+# Velocity engine - probably not needed
```
Should this just come out?

jonrandahl replied:
Sadly no, this is a copy of the conf from the SR Manager, therefore not our code to change

joescottdave reviewed (approved):
LGTM

jonrandahl committed:
fix: Adds `HOTFIX` branch name check in post-commit hook

- Corrects the regex used for checking branch names in the post-commit hook
 - Ensures that hotfix branches are also excluded from checks, in addition to rebase and production branches.

jonrandahl committed:
feat: Enhances request logging information

Improves request logging to include contextual data.

 - Provides more comprehensive request information for debugging and monitoring.
 - Includes request method, parameters, path, and response status in default log output.

joescottdave reviewed (approved):
👍
